fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/lao/event/rollcall/RollCallViewModel.java

```diff
@@ -39,6 +39,7 @@ public class RollCallViewModel extends AndroidViewModel implements QRCodeScannin
   private String laoId;
 
   private final Set<PublicKey> attendees = new HashSet<>();
+  private Runnable adapterUpdate = null;
```

Comment:
Yes, in fact for the list of attendees I simply take the information from the rollcall. However, when displaying the list of scanned tokens for the organizer I need a way to access to the view model. As you suggested, UI logic shouldn't be handled by the view model, so I thought that by using a Runnable I was abstracting the interaction and keeping the two distinct. Now I understand that this is not a good architectural pattern. However, there's a problem since when the view model scans the participants I need a way to trigger the update of the adapter, without the view model to intervene. I'm going to think more about that to come up with a solution

Reply:
I'm really unsure about the what you mean by 

> when displaying the list of scanned tokens for the organizer I need a way to access to the view model

As accessing the viewModel from a Fragment is very common (and you've done it already).
I suspect that you are overcomplexifying things in your head (it's happened to all of us 😄)

Also perhaps it comes from the fact that you haven't realised that the displaying part is different from the scanning part (they are different fragment) and therefore the displaying part needs not caring about newly scanned keys. 

For your information if that was the case, a correct (at least in Android) way of doing things would be to have a `MutableLiveData` (thereafter "mld") in the viewmodel that keeps track of the list of scanned keys. That mld would be updated by the scanner and observed by the scanned keys displaying part. 
But as I said, such a construction is not needed here

Reply:
Yeah, sure about accessing the ViewModel from the Fragment since my current solution was based on that. 

However my point was based on the Adapter, which is contained in the Fragment. From what I understood, in order to work properly, when new data is added to the list which the adapter is attached to, as it's stored in the `attendees` set of the ViewModel, a notifyDataChanged should be called on the adapter, else the list is not updated on the fly, but rather when the function `setUpStateDependantContent` is called again. But if the ViewModel cannot access directly the adapter, and the Fragment has no system of notification on the fact that new attendees are added by the viewModel, then how could the adapter be notified? 

The solution I was thinking was of an Observable on that `attendees` set, so that when it's modified the Fragment can trigger the `notifyDataSetChanged` on the adapter. I don't know if it's overkill and there's a more straightforward solution. Hope it makes sense!

Reply:
Sorry I'm not being clear enough. Your solution works and it would be the usual way to handle things (the `MutableLiveData` I suggested is a simpler way of doing observable in android). 
Yet the thing I pointed to is that this situation is particular in the sense it's simpler. There is no need to dynamically update the list of scanned keys. The only way to add a scanned key is to open the scanner (and the key list won't be displayed i.e. no need to dynamically update the display). Conversely, when displaying the list, no key can be added as only the organizer may scan (and the scanner won't be open). 
Therefore when navigating to the `RollCallFragment` (as an organizer and when it's open), you only need to fetch the scanned list (via the view model), display it and that's it.

Does that make sense ? Otherwise we'll discuss IRL

Reply:
Yes, it's clear now! I just edited the comment realizing what you told me few seconds before your comment, perfect timing 😄